commit ffaf72b9ae058f26e0d0f52f530476d6d7c7c78e
Author: ccat3z <fzhang.chn@outlook.com>
Date:   Mon Jun 26 14:56:01 2023 +0000

    Fix add_library(SHARED) in add_flow_target

diff --git a/cmake/FlowCommands.cmake b/cmake/FlowCommands.cmake
index 01616cd18..e184c15e0 100644
--- a/cmake/FlowCommands.cmake
+++ b/cmake/FlowCommands.cmake
@@ -213,7 +213,7 @@ function(add_flow_target)
         message(FATAL_ERROR "add_flow_target can only be of one type")
       endif()
       set(strip_target ON)
-      add_library(${AFT_NAME} DYNAMIC ${sources} ${AFT_ADDL_SRCS})
+      add_library(${AFT_NAME} SHARED ${sources} ${AFT_ADDL_SRCS})
     endif()
 
     foreach(src IN LISTS sources AFT_ADDL_SRCS)
