diff --git a/bindings/c/CMakeLists.txt b/bindings/c/CMakeLists.txt
index b1a187b99..c3b07d787 100644
--- a/bindings/c/CMakeLists.txt
+++ b/bindings/c/CMakeLists.txt
@@ -1,5 +1,6 @@
 set(FDB_C_SRCS
   fdb_c.cpp
+  fdb_c.actor.cpp
   foundationdb/fdb_c.h
   foundationdb/fdb_c_internal.h
   foundationdb/fdb_c_types.h)
@@ -46,8 +47,7 @@ include(GenerateExportHeader)
 if(OPEN_FOR_IDE)
   add_library(fdb_c OBJECT ${FDB_C_SRCS} ${asm_file})
 else()
-  add_library(fdb_c SHARED ${FDB_C_SRCS} ${asm_file})
-  strip_debug_symbols(fdb_c)
+  add_flow_target(DYNAMIC_LIBRARY NAME fdb_c SRCS ${FDB_C_SRCS} ADDL_SRCS ${asm_file})
 endif()
 add_dependencies(fdb_c fdb_c_generated fdb_c_options)
 add_dependencies(fdbclient fdb_c_options)
